pepper-apis/src/main/java/org/broadinstitute/ddp/elastic/participantslookup/ESParticipantsStructuredIndexSearchHelper.java

```diff
@@ -0,0 +1,131 @@
+import static org.broadinstitute.ddp.elastic.ElasticSearchQueryUtil.getJsonElement;
+import static org.broadinstitute.ddp.elastic.ElasticSearchQueryUtil.prepareSearching;
+import static org.broadinstitute.ddp.elastic.participantslookup.model.ESParticipantsLookupField.isQueryFieldForIndex;
+import static org.broadinstitute.ddp.elastic.participantslookup.model.ESParticipantsLookupIndexType.PARTICIPANTS_STRUCTURED;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.gson.Gson;
+import org.broadinstitute.ddp.elastic.ElasticSearchIndexType;
+import org.broadinstitute.ddp.elastic.participantslookup.model.ESParticipantsLookupField;
+import org.broadinstitute.ddp.elastic.participantslookup.model.ESParticipantsStructuredIndexResultRow;
+import org.broadinstitute.ddp.model.user.EnrollmentStatusType;
+import org.broadinstitute.ddp.service.participantslookup.ParticipantsLookupResult;
+import org.broadinstitute.ddp.util.ElasticsearchServiceUtil;
+import org.broadinstitute.ddp.util.GsonUtil;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.client.RestHighLevelClient;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.Operator;
+import org.elasticsearch.index.query.QueryBuilders;
+
+
+/**
+ * Provides data reading from ES index "participants_structured"
+ */
+public class ESParticipantsStructuredIndexSearchHelper {
+
+    private static final Gson gson = GsonUtil.standardGson();
+
+    private static final String SOURCE__STATUS = "status";
+    private static final String SOURCE__PROFILE = "profile";
+    private static final String SOURCE__INVITATIONS = "invitations";
+    private static final String SOURCE__PROXIES = "proxies";
+
+    /**
+     * ElasticSearch "_source" names (to fetch data from index "participants_structured")
+     */
+    private static final String[] PARTICIPANTS_STRUCTURED__INDEX__SOURCE = {
+            SOURCE__STATUS,
+            SOURCE__PROFILE,
+            SOURCE__INVITATIONS,
+            SOURCE__PROXIES
+    };
+
+    // Field value (to fetch from source 'invitations'
+    private static final String GUID = "guid";
+
+
+    public static Map<String, ESParticipantsStructuredIndexResultRow> searchInParticipantsStructuredIndex(
+            RestHighLevelClient esClient,
+            String studyGuid,
+            String mainQuery,
+            Map<String, String> governedUserToProxy,
+            int resultMaxCount,
+            ParticipantsLookupResult participantsLookupResult) throws IOException {
+        var esIndex = ElasticsearchServiceUtil.detectEsIndex(studyGuid, ElasticSearchIndexType.PARTICIPANTS_STRUCTURED);
+        var queryBuilder = createESQueryForParticipantsStructuredIndex(mainQuery, governedUserToProxy);
+        var searchRequest = prepareSearching(esIndex, PARTICIPANTS_STRUCTURED__INDEX__SOURCE, queryBuilder, resultMaxCount);
```

Comment:
Now that I'm looking at this more closely, I think there's some flaws with our approach.

One concern is this second query and the `resultMaxCount` limit. What if there's more than 100 matches, and ES doesn't return all the governed users that we're looking for? Then we won't be able to build the correct result set.

Another concern is also related to governed users. With our approach, it's possible that this second query returns a governed participant. It's possible that we don't have a proxy for them in our list from our first query since our first query only focused on matching on proxy data.

I was hoping we can avoid doing multiple search queries. But maybe it's unavoidable. Perhaps we need 4 queries:

* First, search proxies in `users` index. Second, lookup their corresponding governed users in `participants_structured` index (this second request is not a search but rather just looking up by `guid`).
* Third, search participants in `participants_structured` index. Fourth, lookup their proxy in the `users` index (again, lookup by `guid` instead of a search).

Reply:
I agree with it.
Redesigned according to the suggested logic.